march_hardware/src/EtherCAT/EthercatMaster.cpp

```diff
@@ -162,6 +170,9 @@ void EthercatMaster::ethercatLoop()
     }
     else
     {
+      std::lock_guard<std::mutex> lock(this->wait_on_pdo_condition_mutex_);
+      this->pdo_received_ = true;
+      this->wait_on_pdo_condition_var_.notify_one();
```

Comment:
Why only notify when a cycle was achieved? It could be that the cycle took longer but still received its PDOs, due to OS scheduling. I would let `sendReceivePdo` return wether the pdos were received and only then notify the cv.

And small note, you have to unlock the mutex before [`notify_one`](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) or the cv will wake up and immediately go back to sleep.
```suggestion
      std::lock_guard<std::mutex> lock(this->wait_on_pdo_condition_mutex_);
      this->pdo_received_ = true;
      lock.unlock();
      this->wait_on_pdo_condition_var_.notify_one();
```

Reply:
Hmm, is this unnecessary in the example because it automatically gets destroyed when it exists {}?
![image](https://user-images.githubusercontent.com/47626418/77416243-f2eead80-6dc3-11ea-81ef-5c1a4f699198.png)

Reply:
Yes

Reply:
I now only send a signal if the wkc is 29. Probably not ideal since you stop updating the controller when you loose a slave, but we saw in the logs that sometimes only a part of the slaves send data, probably due to wkc too low. Ideally, of course, you would like to then update only those slaves from whom you received data. That is a bit complicated in the current format, however :/.

Reply:
Hmm this is getting a bit complicated, since you want to keep updating all slaves. Say for example that every cycle another slave loses connection, first 1 then 2,3,... Then no slave gets updated :frowning_face:

Reply:
Perhaps we have to go for/combine this with another approach after all. I can create a  check within the joint class whether the parameters have been updated (for example by assuming the the motor voltage never completely remains the same). I can then call this check for every joint separately in the loop of MarchHardwareInterface::read.

Reply:
Okay, let's close this PR for now